src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -189,168 +248,198 @@ public long ramBytesUsed() {
                         }
                     }
                 }
             }
-        } catch (IOException ex) {
-            logger.debug("failed to clear unreferenced files ", ex);
-        } finally {
-            rwl.writeLock().unlock();
         }
-        return deleted;
+        return result.toArray(Strings.EMPTY_ARRAY);
     }
 
     @Override
-    public void newTranslog(long id) throws TranslogException, IOException {
-        rwl.writeLock().lock();
-        try {
-            FsTranslogFile newFile;
-            long size = Long.MAX_VALUE;
-            Path location = null;
-            for (Path file : locations) {
-                long currentFree = Files.getFileStore(file).getUsableSpace();
-                if (currentFree < size) {
-                    size = currentFree;
-                    location = file;
-                } else if (currentFree == size && ThreadLocalRandom.current().nextBoolean()) {
-                    location = file;
+    public void committedTranslogId(final long translogId) throws FileNotFoundException {
+        try (ReleasableLock lock = writeLock.acquire()) {
+            logger.trace("updating translogs on commit of [{}]", translogId);
+            if (translogId < lastCommittedTranslogId) {
+                throw new ElasticsearchIllegalArgumentException("committed translog id can only go up (current ["
+                        + lastCommittedTranslogId + "], got [" + translogId + "]");
+            }
+            boolean found = false;
+            if (current.translogId() == translogId) {
+                found = true;
+            } else {
+                if (translogId > current.translogId()) {
+                    throw new ElasticsearchIllegalArgumentException("committed translog id must be lower or equal to current id (current ["
+                            + current.translogId() + "], got [" + translogId + "]");
                 }
             }
-            try {
-                newFile = type.create(shardId, id, new InternalChannelReference(location.resolve(getPath(id)), StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW), bufferSize);
-            } catch (IOException e) {
-                throw new TranslogException(shardId, "failed to create new translog file", e);
+            if (found == false) {
+                // try to find it in uncommittedTranslogs
+                for (int i = uncommittedTranslogs.size() - 1; i >= 0; i--) {
+                    if (uncommittedTranslogs.get(i).translogId() == translogId) {
+                        found = true;
+                        break;
+                    }
+                }
             }
-            FsTranslogFile old = current;
-            current = newFile;
-            IOUtils.close(old);
-        } finally {
-            rwl.writeLock().unlock();
-        }
-    }
+            if (found == false) {
+                ArrayList<Long> currentIds = new ArrayList<>();
```

Comment:
can you just use `Iterables#concat(uncommittedTranslogs, Collections.singletonList(current))`